backend/rs/ic_vetkeys/src/utils/mod.rs

```
@@ -397,14 +397,98 @@ impl EncryptedVetKey {
     }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+/// An identity, used for identity based encryption
+///
+/// As far as the IBE encryption scheme goes this is simply an opauqe bytestring
+/// We provide a type to make code using the IBE a bit easier to understand
+pub struct Identity {
+    val: Vec<u8>,
+}
+
+impl Identity {
+    /// Create an identity from an arbitrary byte string
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Self {
+            val: bytes.to_vec(),
+        }
+    }
+
+    /// Create an identity from an arbitrary UTF8 string
+    pub fn from_string(str: &str) -> Self {
+        Self::from_bytes(str.as_bytes())
+    }
+
+    /// Create an identity from a Principal
+    pub fn from_principal(principal: &candid::Principal) -> Self {
+        Self::from_bytes(principal.as_slice())
+    }
+
+    /// Return the bytestring of this identity
+    pub fn value(&self) -> &[u8] {
+        &self.val
+    }
+}
+
+/*
+* Amount of randomness generated during the IBE encryption process
+*/
 const IBE_SEED_BYTES: usize = 32;
 
+/// A random seed, used for identity based encryption
+pub struct Seed {
+    val: [u8; IBE_SEED_BYTES],
+}
+
+impl Seed {
+    /// Create a random seed for IBE encryption
+    pub fn random<R: rand::CryptoRng + rand::RngCore>(rng: &mut R) -> Self {
+        use rand::Rng;
+        Self {
+            val: rng.gen::<[u8; IBE_SEED_BYTES]>(),
+        }
+    }
+
+    /// Create a seed for IBE encryption from a byte string
+    ///
+    /// This input should be randomly chosen by a secure random number generator.
+    /// If the seed is not securely generated the IBE scheme will be insecure.
+    ///
+    /// If the input is exactly 256 bits it is used directly. Otherwise the input
```

Do we want to return an error if the input is too small?

---

I had wanted to make it infalliable but ignoring eg empty input is probably a footgun. I changed it to require at least 128 bits since that ~~ the security level of BLS12-381 anyway. Hopefully everyone will just use `IbeSeed::random` instead